api/src/services/authentication.ts

```diff
@@ -429,6 +438,63 @@ export class AuthenticationService {
 		};
 	}
 
+	private async updateStatefulSession(
+		sessionRecord: Record<string, any>,
+		oldSessionToken: string,
+		newSessionToken: string,
+		sessionExpiration: Date,
+	): Promise<string> {
+		if (sessionRecord['session_next_token']) {
+			// The current session token was already refreshed and has a reference
+			// to the new session, update the new session timeout for the new refresh
+			await this.knex('directus_sessions')
+				.update({
+					expires: sessionExpiration,
+				})
+				.where({ token: newSessionToken });
+
+			return newSessionToken;
+		}
+
+		// Keep the old session active for a short period of time
+		const GRACE_PERIOD = getMilliseconds(env['SESSION_REFRESH_GRACE_PERIOD'], 10_000);
+
+		// Update the existing session record to have a short safety timeout
+		// before expiring, and add the reference to the new session token
+		const updatedSession = await this.knex('directus_sessions')
+			.update(
+				{
+					next_token: newSessionToken,
+					expires: new Date(Date.now() + GRACE_PERIOD),
+				},
+				['next_token'],
+			)
+			.where({ token: oldSessionToken, next_token: null });
+
+		if (updatedSession.length === 0) {
+			// Don't create a new session record, we already have a "next_token" reference
+			const { next_token } = await this.knex('directus_sessions')
+				.select('next_token')
+				.where({ token: oldSessionToken })
+				.first();
+
+			return next_token;
+		}
```

Comment:
Would be perfect if we could combine that into one query, but I'm afraid there's no generic/easy way to do so cross vendor.

Example for Postgres:
```sql
UPDATE directus_sessions
SET    next_token =
       CASE
              WHEN next_token IS NULL THEN 'newSessionToken'
              ELSE next_token
       END
WHERE  token = 'oldSessionToken'
RETURNING next_token; 
```

Reply:
Played around with that as well, didn't arrive at a good solution that's supported across vendors. But we also need to keep in mind that that race condition where we actually do the select query is super slim and as such the amount of duplicates queries should be super small over all.

Reply:
Agreed 👍